components/application-broker/internal/storage/populator/instances.go

```diff
@@ -69,11 +93,132 @@ func (p *Instances) Do(ctx context.Context) error {
+		}
+		addInstance = true
+	case provisionPollingLastOperationMode:
+		p.log.Info("ServiceInstance is in failed provision state with PollingLastOperation error")
+		var params map[string]interface{}
+		if si.Spec.Parameters != nil {
+			err := json.Unmarshal(si.Spec.Parameters.Raw, &params)
+			if err != nil {
+				return errors.Wrap(err, "while unmarshaling instance parameters")
+			}
+		}
+		p.log.Info("restore provisioning process")
+		err := p.broker.ProvisionProcess(broker.RestoreProvisionRequest{
+			Parameters:           params,
+			InstanceID:           internal.InstanceID(si.Spec.ExternalID),
+			OperationID:          internal.OperationID(*si.Status.LastOperation),
+			Namespace:            internal.Namespace(si.Namespace),
+			ApplicationServiceID: internal.ApplicationServiceID(si.Spec.ServiceClassRef.Name),
+		})
+		if err != nil {
+			return errors.Wrap(err, "while triggering provisioning process")
+		}
+
+		_, err = p.addOperation(si, internal.OperationTypeCreate, internal.OperationStateInProgress, internal.OperationID(*si.Status.LastOperation))
+		if err != nil {
+			return errors.Wrap(err, "while add operation for failed instance (PollingLastOperation error)")
+		}
+		addInstance = true
+	case deprovisionPollingLastOperationMode, deprovisionCallFailedMode:
+		p.log.Info("ServiceInstance is in failed deprovision state")
+		_, err := p.addOperation(si, internal.OperationTypeCreate, internal.OperationStateSucceeded, "")
+		if err != nil {
+			return errors.Wrap(err, "while add operation for failed instance (DeprovisionCallFailed)")
+		}
+		opID, err := p.addOperation(si, internal.OperationTypeRemove, internal.OperationStateInProgress, "")
+		if err != nil {
+			return errors.Wrap(err, "while add operation for failed instance (DeprovisionCallFailed)")
+		}
+		p.broker.DeprovisionProcess(broker.DeprovisionProcessRequest{
+			Instance:             p.converter.MapServiceInstance(si),
+			OperationID:          opID,
+			ApplicationServiceID: internal.ApplicationServiceID(si.Spec.ServiceClassRef.Name),
+		})
+	default:
+		p.log.Infof("ServiceInstance will to be populated, unsupported mode (%s - last state: %s)",
+			si.Status.CurrentOperation,
+			si.Status.LastConditionState)
+	}
+
+	if !addInstance {
+		return nil
+	}
+
+	if err := p.inserter.Insert(p.converter.MapServiceInstance(si)); err != nil {
+		return errors.Wrap(err, "while inserting service instance")
+	}
+
+	return nil
+}
+
+func (p *Instances) specifyRestoreMode(instance *v1beta1.ServiceInstance) string {
+	for _, cond := range instance.Status.Conditions {
+		if cond.Type == v1beta1.ServiceInstanceConditionReady && cond.Status == v1beta1.ConditionTrue {
+			return readyMode
+		}
+	}
+
+	// case when AppBroker restarts before SC receives response with operation ID
+	if instance.Status.LastConditionState == "DeprovisionCallFailed" {
+		return deprovisionCallFailedMode
+	}
+
+	if instance.Status.LastOperation == nil {
```

Comment:
shouldn't we check this at beginning?

Reply:
in the beginning, instances with ready status are checked where `LastOperation` is not necessary, then instances with failed status which didn't get a response so it means last operation id was not responded. After that instances with failed status and `ErrorPollingLastOperation` reason are checked what means they got a response and wait for operation status.